cosmovisor/upgrade.go

```diff
@@ -69,16 +69,64 @@ func DownloadBinary(cfg *Config, info *UpgradeInfo) error {
 		// copy binary to binPath from dirPath if zipped directory don't contain bin directory to wrap the binary
 		if err != nil {
 			err = copy.Copy(filepath.Join(dirPath, cfg.Name), binPath)
+		}
+		if err != nil {
+			// If there is only one entry in dirPath, then promote it to the upgradedir
+
+			// Clean up an empty bin directory.
+			os.Remove(filepath.Join(dirPath, "bin"))
+
+			var childName string
+			childName, err = GetOnlyChildName(dirPath)
 			if err != nil {
 				return err
 			}
+
+			// move aside the old dir and move the child to the old dir
+			var tmpPath string
+			tmpPath, err = os.MkdirTemp(filepath.Join(dirPath, ".."), "tmp*")
+			if err != nil {
+				return err
+			}
+			asidePath := filepath.Join(tmpPath, "old")
+			err = os.Rename(dirPath, asidePath)
+			if err == nil {
+				err = os.Rename(filepath.Join(asidePath, childName), dirPath)
```

Comment:
Do you need to retry `copy.Copy(filepath.Join(dirPath, cfg.Name), binPath)` here?